feat: add omnipool math

enthusiastmartin opened the pull request.

enthusiastmartin committed:
feat: add omnipool math

enthusiastmartin committed:
use version 16 of substrate version for now

enthusiastmartin committed:
pin fixed crate version

enthusiastmartin committed:
use sp_std

enthusiastmartin committed:
remove simple imbalance

enthusiastmartin committed:
use checked math

enthusiastmartin committed:
omnipool add invariants testS

enthusiastmartin committed:
use checked math

mrq1911 commented on Cargo.toml:
```diff
@@ -6,15 +6,17 @@ license = 'Apache-2.0'
 # Added because of use of the FixedU128
 codec = { package = "parity-scale-codec", version = "2.3.1", default-features = false, features = ["derive", "max-encoded-len"] }
-sp-arithmetic = { git = "https://github.com/paritytech/substrate", rev = "22d40c761a985482f93bbbea5ba4199bdba74f8e", version="4.0.0", default-features = false }
+#sp-arithmetic = { git = "https://github.com/paritytech/substrate", rev = "22d40c761a985482f93bbbea5ba4199bdba74f8e", version="4.0.0", default-features = false }
```
remove

mrq1911 commented on src/omnipool/types.rs:
```diff
@@ -0,0 +1,365 @@
+impl<Balance: CheckedAdd + CheckedSub + PartialOrd + Default> Add<Self> for BalanceUpdate<Balance> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
```
should we even have implementation for unsafe math?

enthusiastmartin replied:
it is required if you implement CheckedAdd.

mrq1911 replied:
make sense :)

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+use sp_arithmetic::{FixedU128, Permill};
+
+pub const ONE: Balance = 1_000_000_000_000;
+pub const TOLERANCE: Balance = 1_000; // * 1_000 * 1_000;
```
what do you want to reveal with this comment?

1_000 * 1_000 * 1_000 = 1_000_000_000, which is ... ?

dmoka commented on src/omnipool/types.rs:
```diff
@@ -0,0 +1,365 @@
+    //use cool_asserts::assert_panics;
+
+    #[test]
+    fn balance_update_addition_works() {
```
maybe you could consider using the [test-case](https://crates.io/crates/test-case) crate, so this test could be simplified, resulting in much less test code.

If you would go with that, then it could be applied in the next tests of this file.

dmoka commented on src/omnipool/types.rs:
```diff
@@ -0,0 +1,365 @@
+            BalanceUpdate::Increase(u128::MAX) + BalanceUpdate::Decrease(1),
+            BalanceUpdate::Increase(u128::MAX - 1)
+        );
+
+        //assert_panics!(BalanceUpdate::Increase(u128::MAX) + BalanceUpdate::Increase(1));
+        //assert_panics!(BalanceUpdate::Decrease(u128::MAX) + BalanceUpdate::Decrease(1));
```
please enable/remove these two lines accordingly

lumir-mrkva replied:
```suggestion
```

mrq1911 replied:
```suggestion
```

dmoka commented on src/omnipool/types.rs:
````diff
@@ -0,0 +1,365 @@
+///
+/// # Example
+///
+/// ```ignore
````
what is the reason to ignore this doc test? Or it would not work due to some dependency?

Because I think it would be handy to have an executable doc test, if possible.

dmoka commented on src/omnipool/math.rs:
```diff
@@ -0,0 +1,425 @@
+    })
+}
+
+pub fn calculate_asset_tvl(asset_hub_reserve: Balance, stable_asset: (Balance, Balance)) -> Option<Balance> {
```
I think that we could add tests for both this `calculate_asset_tvl` and  `calculate_delta_imbalance`, as they are parts of the public API.

dmoka commented on src/omnipool/math.rs:
```diff
@@ -0,0 +1,425 @@
+    })
+}
+
+/// Calculate delta changes of rmove liqudiity given current asset state and position from which liquidity should be removed.
```
```suggestion
/// Calculate delta changes of remove liquidity given the current asset state and position from which liquidity should be removed.
```

dmoka commented on Cargo.toml:
```diff
@@ -6,15 +6,17 @@ license = 'Apache-2.0'
 primitive-types = {default-features = false, version = '0.8.0'}
 num-traits = {default-features = false, version = "0.2.14"}
-fixed = "1.9.0"
+fixed = "=1.15.0"
```
is this `=` necessary?

enthusiastmartin replied:
yes.otherwise it updated to 1.16 which does not work .

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn add_liquidity_prices(asset in asset_state(),
```
```suggestion
    fn price_should_not_change_when_liquidity_added(asset in asset_state(),
```

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn remove_liquidity_prices(asset in asset_state(),
```
```suggestion
    fn price_should_not_change_when_liquidity_removed(asset in asset_state(),
```

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+}
+#[test]
+fn buy_update_invariants_no_fees_case() {
+    let asset_in = AssetReserveState {
```
I think we could also use the number-representation-with-underscore for property values of this struct, so we follow the styling convention.

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+}
+
+fn trade_amount() -> impl Strategy<Value = Balance> {
+    //Just(1000 * ONE)
```
is this a leftover comment, or needed?

mrq1911 replied:
```suggestion
```

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+            Balance::default()
+        );
+
+        // ignore the invalid result
```
I think that instead of telling what we do not do, would be better to tell what we do

```suggestion
        // perform assertion only when result is valid
```

If you go for this, please do it in the previous test as well.

dmoka commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,323 @@
+        let new_asset_state = new_asset_state.delta_update(&state_changes.asset).unwrap();
+
+        // Price should not change
+        assert_eq_approx!(asset.price().unwrap(),
```
I would extract it to a helper method called `assert_that_price_has_not_changed`, then also using it in the next test.

By doing so, we can also get rid of the comment.

The 3rd param of this approx method can be more generic, and if the test fails, we can easily get the corresponding reason if you do my proposed test renamings.

dmoka commented on src/omnipool/math.rs:
```diff
@@ -0,0 +1,425 @@
+        .and_then(|v| v.checked_div(position_shares_hp))?;
+
+    let adjusted_asset_tvl_hp = if is_stable_asset {
+        // TODO: add test for this
```
todo :)

enthusiastmartin committed:
comments

poliwop committed:
Fixing rounding issues in withdraw liquidity calculations, relating to protocol owned shares and LRNA dispersal.
Adding property tests to catchs such rounding issues in the future.

poliwop committed:
Added a macro assert_eq_approx_ordered which, in addition to panicing if the two provided values are not within a provided tolerance, also panics if the first value is strictly less than the second value.
Switched the add/remove liquidity tests to use the new macro.

enthusiastmartin committed:
refactor

enthusiastmartin committed:
use safe tou128

mrq1911 committed:
Apply suggestions from code review

Co-authored-by: Daniel Moka <mokabme@gmail.com>

mrq1911 committed:
cleanup

mrq1911 commented on src/omnipool/invariants.rs:
```diff
@@ -0,0 +1,322 @@
+#[test]
+fn buy_update_invariants_no_fees_case() {
+    let asset_in = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 10000000000000000,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
+    };
+    let asset_out = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 89999999999999991,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
```
```suggestion
        reserve: 10_000_000_000_000_000,
        hub_reserve: 10_000_000_000_000_000,
        shares: 10_000_000_000_000_000,
        protocol_shares: 10_000_000_000_000_000,
        tvl: 10_000_000_000_000_000,
    };
    let asset_out = AssetReserveState {
        reserve: 10_000_000_000_000_000,
        hub_reserve: 89_999_999_999_999_991,
        shares: 10_000_000_000_000_000,
        protocol_shares: 10_000_000_000_000_000,
        tvl: 10_000_000_000_000_000,
```

mrq1911 committed:
_ _ _ _ _

mrq1911 committed:
Merge branch 'main' into feat/omnipool

mrq1911 commented:
@enthusiastmartin please resolve @dmoka comments

enthusiastmartin committed:
Merge branch 'main' into feat/omnipool

enthusiastmartin committed:
enable doc test

enthusiastmartin committed:
use test-case to simplify test code

enthusiastmartin committed:
refactor: use already calculated value

enthusiastmartin committed:
clippy ?

enthusiastmartin committed:
Merge pull request #54 from galacticcouncil/refactor/omnipool-51

refactor: omnopool - use already calculated value

enthusiastmartin committed:
use polkadot 16 for omnipool math

enthusiastmartin commented:
> @enthusiastmartin please resolve @dmoka comments

we can't really merge this just yet.

the reason being that HydraDX-node is on polkadot-v0.9.16 and Basilisk on v17. 

Merging this PR would downgrade to polkadot v16 because this version is needed to be compatible with current omnipool pallet version which uses v16.

enthusiastmartin committed:
add simple tests for all omnipool math

enthusiastmartin committed:
Merge branch 'main' into feat/omnipool

enthusiastmartin committed:
happy clippy happy life

enthusiastmartin committed:
formatting

enthusiastmartin committed:
Merge branch 'feat/omnipool' into feat/omnipool-rounding-fix

enthusiastmartin committed:
add additional remove liquid calc tests

enthusiastmartin committed:
Merge branch 'feat/omnipool' into feat/omnipool-rounding-fix

enthusiastmartin committed:
fix tests to reflect rounding changes

enthusiastmartin committed:
fix: rounding when computing amount without fee

enthusiastmartin committed:
refactor: imbalance update in add/remove liquidity

enthusiastmartin committed:
fix imbalance upadte in lrna swaps

enthusiastmartin committed:
fix correct imbalance update in lrna swaps

enthusiastmartin committed:
Merge branch 'feat/omnipool-rounding-fix' into audit-fixes

enthusiastmartin committed:
Merge branch 'refactor/imbalance-update' into audit-fixes

enthusiastmartin committed:
ajdust test due to rounding changes

enthusiastmartin committed:
Merge branch 'fix/imbalance-lrna' into audit-fixes

enthusiastmartin committed:
add calculate_tvl

enthusiastmartin committed:
remove tvl

enthusiastmartin committed:
upadte invariants tests

enthusiastmartin committed:
rework imbalance update

enthusiastmartin committed:
remove not used stuff

enthusiastmartin committed:
happy clippy happy life

enthusiastmartin committed:
Merge pull request #62 from galacticcouncil/refactor/omnipool-tvl

refactor: omnipool - remove asset tvl calculations

enthusiastmartin committed:
formatting

enthusiastmartin committed:
remove obsolete delta imbalance calculation

enthusiastmartin committed:
fix imbalance rounding in lrna swaps

enthusiastmartin committed:
reduce duplicate code

enthusiastmartin committed:
use already calculate value

poliwop committed:
WIP (some arithmatic in this update needs to be made safe)

Added an invariant test demonstrating that the old rounding was not passing the desired tests, for sufficiently large (but possible) values of inputs.
Tweaked the rounding to pass the tests.

enthusiastmartin committed:
use safe math

enthusiastmartin committed:
fix: omnipool rounding adjustments

enthusiastmartin committed:
formatting

enthusiastmartin committed:
remove redundant precondition check

enthusiastmartin committed:
Merge pull request #70 from galacticcouncil/omnipool-final-fixes

fix: omnipool final rounding adjustments

enthusiastmartin committed:
Merge pull request #61 from galacticcouncil/audit-fixes

fix: audit issues

dmoka commented on src/omnipool/types.rs:
```diff
@@ -204,72 +204,45 @@ mod tests {
+    #[test_case(BalanceUpdate::Decrease(0), BalanceUpdate::Decrease(100), BalanceUpdate::Decrease(100) ; "When decrease zero and decreaes ")]
+    #[test_case(BalanceUpdate::Increase(100), BalanceUpdate::Decrease(100), BalanceUpdate::Increase(0) ; "When decrease and decrease same amount ")]
+    #[test_case(BalanceUpdate::Decrease(100), BalanceUpdate::Increase(100), BalanceUpdate::Decrease(0) ; "When decrease and decrease same amount swapped ")] // should be probably same as previous ?
+    #[test_case(BalanceUpdate::Increase(u32::MAX), BalanceUpdate::Decrease(1), BalanceUpdate::Increase(u32::MAX - 1) ; "When increase max and decrease one")]
```
awesome test set up, great job!

enthusiastmartin committed:
Merge branch 'main' into feat/omnipool

enthusiastmartin committed:
fix transcenadal tests

enthusiastmartin committed:
formatting

enthusiastmartin committed:
happy clippy happy life
